bbq/compiler/compiler.go

```diff
@@ -577,17 +601,75 @@ func (c *Compiler[_, _]) compileDeclaration(declaration ast.Declaration) {
 	ast.AcceptDeclaration[struct{}](declaration, c)
 }
 
-func (c *Compiler[_, _]) compileBlock(block *ast.Block) {
+func (c *Compiler[_, _]) compileBlock(block *ast.Block, enclosingDeclKind common.DeclarationKind) {
 	locals := c.currentFunction.locals
 	locals.PushNewWithCurrent()
 	defer locals.Pop()
 
-	for _, statement := range block.Statements {
+	// Functions (regular functions and initializers) can have post conditions.
+	switch enclosingDeclKind {
+	case common.DeclarationKindFunction, common.DeclarationKindInitializer:
+		if c.hasPostConditions() {
+			c.pushReturns()
+		}
+	}
+
+	for index, statement := range block.Statements {
+		// Once the post conditions are reached, patch all the previous return statements
+		// to jump to the current index (i.e: update them to jump to the post conditions).
+		if index == c.postConditionsIndex {
```

Comment:
How are functions with nested blocks handled? `compileFunctionBlock` sets the `postConditionsIndex` field, but then here in `compileBlock` it seems to be checked for all blocks of the function.

Reply:
That's a good point! It should only patch the jumps at the function-block level, but not at the nested block levels. Currently it does both, and the same jump get patched twice (incorrectly), but since the outer-most patch is always the function-block (a "sandwich"), it ends-up producing the correct jump-index eventually. 

Fixed it in https://github.com/onflow/cadence/pull/3816/commits/e9b789ea18187043dc63f0dc37eda4f8eb5316dc, to only patch the returns for function-blocks, but not for nested/non-function blocks.
However, it's not actually test-able, since the end result is the same.

Great catch!